functions.lua

```diff
@@ -250,6 +254,166 @@ function PacifistMod.remove_misc()
         assert(entry[1] and entry[2])
         data_raw.remove(entry[1], entry[2])
     end
+
+    -- hide explosion entities revealed by removing/hiding other things
+    data_raw.hide("explosion", "atomic-nuke-shockwave")
+    data_raw.hide("explosion", "wall-damaged-explosion")
+end
+
+function PacifistMod.record_references()
+    -- When we see a name, we aren't being careful enough to know what type
+    -- that name is, so if multiple types have the same name, treat a reference
+    -- to any one of them as a reference to all of them. This may lead to not
+    -- removing something that's actually unreferenced, but won't ever result in
+    -- accidentally removing something this is referenced.
+
+    -- all_names is a map from names to all types that have that name.
+    local all_names = {}
+
+    -- references[group][name]['from'][other_group][other_name]
+    -- is true (not nil) if there is a reference from data.raw[group][name]
+    -- to data.raw[other_group][other_name]. There should be a matching entry
+    -- references[other_group][other_name]['to'][group][name] which indicates
+    -- data.raw[other_group][other_name] has a reference to it
+    -- from data.raw[group][name].
+    local references = {}
+
+    -- Read just the keys of data.raw to initialize those data structures.
+    for group, list in pairs(data.raw) do
+        local refs_for_group = {}
+        references[group] = refs_for_group
+        for _, entity in pairs(list) do
+            if not all_names[entity.name] then
+                all_names[entity.name] = {}
+            end
+            all_names[entity.name][group] = true
+            refs_for_group[entity.name] = {
+                -- Collection of names referenced from this entity.
+                from = {},
+                -- Collection of names with references to this entity.
+                -- If it's empty, then this entity is unreferenced and can be removed.
+                to = {},
+            }
+        end
+    end
+
+    -- Helper that recurses through an object x which is data.row[from_group][from_name]
```

Comment:
```suggestion
    -- Helper that recurses through an object x which is data.raw[from_group][from_name]
```

Reply:
Oops. I never actually pushed the fix for this typo, so it ended up in `main`.

Reply:
I'll fix it next time I work on it (if I remember)

Reply:
ca3f45d